drivers/sensor/scd30/scd30.c

```diff
@@ -0,0 +1,239 @@
+{
+	uint8_t actual_crc = scd30_compute_crc(data, data_len);
+
+	if (checksum != actual_crc) {
+		LOG_ERR("CRC check failed. Expected: %x, got %x", checksum, actual_crc);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int scd30_write_register(const struct device *dev, uint16_t cmd, uint16_t val)
+{
+	const struct scd30_config *cfg = dev->config;
+	uint8_t tx_buf[SCD30_CMD_SINGLE_WORD_BUF_LEN];
+
+	sys_put_be16(cmd, &tx_buf[0]);
+	sys_put_be16(val, &tx_buf[2]);
+	tx_buf[4] = scd30_compute_crc(tx_buf, sizeof(tx_buf));
+
+	return i2c_write_dt(&cfg->bus, tx_buf, sizeof(tx_buf));
+}
+
+/**
+ * @brief Take the received  word (16 bit), if the data is valid, add it to dst
+ *
+ * @param data data received from sensor
+ * @param dst buffer to store data
+ * @return int
+ */
+static int scd30_fill_data_buf(uint8_t *data, uint8_t *dst)
+{
+	if (scd30_check_crc(data, 2, data[2])) {
+		return -EIO;
+	}
+
+	dst[0] = data[0];
+	dst[1] = data[1];
+
+	return 0;
+}
+
+float scd30_bytes_to_float(const uint8_t *bytes)
+{
+	union {
+		uint32_t u32_value;
+		float float32;
+	} tmp;
+
+	tmp.u32_value = sys_get_be32(bytes);
+	return tmp.float32;
+}
+
+static int scd30_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	uint8_t rx_buf[6 * (SCD30_WORD_SIZE + SCD30_CRC8_LEN)] = {};
+	uint8_t data_buf[6 * SCD30_WORD_SIZE] = {};
+	struct scd30_data *data = dev->data;
+	const struct scd30_config *cfg = dev->config;
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	rc = scd30_write_command(dev, SCD30_CMD_GET_DATA_READY);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = i2c_read_dt(&cfg->bus, rx_buf, SCD30_WORD_SIZE + SCD30_CRC8_LEN);
+
+	rc = scd30_fill_data_buf(rx_buf, data_buf);
+	if (rc != 0) {
+		return rc;
+	}
+
+	uint16_t data_ready = sys_get_be16(data_buf);
+
+	if (!data_ready) {
+		return -ENODATA;
+	}
+
+	rc = scd30_write_command(dev, SCD30_CMD_READ_MEASUREMENT);
+	if (rc != 0) {
+		LOG_DBG("Failed to send command. (rc = %d)", rc);
+		return rc;
+	}
+
+	/* delay for 3 msec as per datasheet. */
+	k_busy_wait(3000);
+
+	rc = i2c_read_dt(&cfg->bus, rx_buf, sizeof(rx_buf));
+	if (rc != 0) {
+		LOG_DBG("Failed to read data. (rc = %d)", rc);
+		return rc;
+	}
+
+	/* C02 data */
+	rc = scd30_fill_data_buf(rx_buf, data_buf);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 3, data_buf + 2);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* temperature data */
+	rc = scd30_fill_data_buf(rx_buf + 6, data_buf + 4);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 9, data_buf + 6);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* relative humidity */
+	rc = scd30_fill_data_buf(rx_buf + 12, data_buf + 8);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 15, data_buf + 10);
+	if (rc != 0) {
+		return rc;
+	}
+
+	data->co2_ppm = scd30_bytes_to_float(data_buf);
+	data->temp = scd30_bytes_to_float(data_buf + 4);
+	data->rel_hum = scd30_bytes_to_float(data_buf + 8);
```

Comment:
There are too many hard-coded offsets here. Can you make `rx_buf` and `data_buf` structures instead?